### PR TITLE
[4.4] Add Support for Python 3.10

### PR DESCRIPTION
**Pull request by Prychantovskyi**
4.4 driver changes for Python 3.10

### CODE REVIEW

File: tests/unit/test_conf.py [hunk 1]
```diff
@@ -165,37 +166,37 @@ def test_pool_config_set_value():
...
-            dict(consumed_pool_config.items())
-        )
-    with _pool_config_deprecations():
-        consumed_pool_config = PoolConfig.consume(
-            dict(consumed_pool_config.items())
-        )
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", DeprecationWarning)
+        consumed_pool_config = PoolConfig.consume(dict(consumed_pool_config.items()))
+        consumed_pool_config = PoolConfig.consume(dict(consumed_pool_config.items()))
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: Why are we muting the `DeprecationWarning`s instead of asserting that they get emitted?

Comment (Prychantovskyi) [reply tests/unit/test_conf.py]: Because in this line Deprecation warning was raised in this lines.

Comment (robsdedude) [reply tests/unit/test_conf.py]: Sure, but we expect them and probably want the tests to fail if they stop being emitted.

File: TESTING.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Neo4j Driver Testing
 
 To run driver tests, [Tox](https://tox.readthedocs.io) is required as well as at least one version of Python.
-The versions of Python supported by this driver are CPython 3.6, 3.7, 3.8, and 3.9.
+The versions of Python supported by this driver are CPython 3.6, 3.7, 3.8, 3.9 and 3.10.
```
Comment (robsdedude) [inline TESTING.md]:
```suggestion
The versions of Python supported by this driver are CPython 3.6, 3.7, 3.8, 3.9, and 3.10.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # Neo4j Driver Change Log
 
+## Version 4.4.9
+
+- Python 3.10 support added (driver version 4.4.9)
```
Comment (robsdedude) [inline CHANGELOG.md]:
```suggestion
- Python 3.10 support added
```

No need to state it twice.

File: tests/unit/test_conf.py [hunk 4]
```diff
@@ -35,6 +37,7 @@
 )
 from neo4j.api import (
     TRUST_SYSTEM_CA_SIGNED_CERTIFICATES,
+    TRUST_ALL_CERTIFICATES,
```
Comment (robsdedude) [inline tests/unit/test_conf.py]:
```suggestion
    TRUST_ALL_CERTIFICATES,
    TRUST_SYSTEM_CA_SIGNED_CERTIFICATES,
```
Please sort imports alphabetically.

File: neo4j/conf.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
 # limitations under the License.
 
 
+import sys
 from abc import ABCMeta
 from collections.abc import Mapping
 import warnings
```
Comment (robsdedude) [inline neo4j/conf.py]:
```suggestion
from abc import ABCMeta
from collections.abc import Mapping
import sys
import warnings
```

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
 from threading import RLock
 
 import pytest
+import warnings
 import urllib
```
Comment (robsdedude) [inline tests/integration/conftest.py]:
```suggestion
import pytest
import urllib
import warnings
```

File: tests/stub/conftest.py [hunk 7]
```diff
@@ -20,13 +20,16 @@
 
 
 import subprocess
+import warnings
 import os
```
Comment (robsdedude) [inline tests/stub/conftest.py]:
```suggestion
import subprocess
import os
import warnings
```

File: tests/unit/test_conf.py [hunk 8]
```diff
@@ -22,6 +22,8 @@
 from contextlib import contextmanager
 import warnings
 
+import ssl
+import sys
 import pytest
```
Comment (robsdedude) [inline tests/unit/test_conf.py]:
```suggestion
from contextlib import contextmanager
import ssl
import sys
import warnings

import pytest
```

There are usually three (or four) blocks of imports (each one alphabetically sorted) separated by a blank line:
 * standard lib packages (those that come shipped with Python)
 * third party packages (those you'd install with pip or similar)
 * local imports (importing package internal things)
 * (only for tests) importing the things under test
 
 The reason for this is:
 * Easier to read and find things if you stick to a consistent scheme
 * Should some import collide with another, you at least make sure that the most important things are last and won't be overwritten.

File: tests/unit/test_conf.py [hunk 9]
```diff
@@ -35,6 +37,7 @@
 )
 from neo4j.api import (
     TRUST_SYSTEM_CA_SIGNED_CERTIFICATES,
+    TRUST_ALL_CERTIFICATES,
     WRITE_ACCESS,
     READ_ACCESS,
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: While not part of you change, feel free to also fix the order here.

File: tests/unit/test_conf.py [hunk 10]
```diff
@@ -93,7 +96,8 @@ def test_pool_config_consume():
 
     test_config = dict(test_pool_config)
 
-    with _pool_config_deprecations():
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", DeprecationWarning)
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: `with _pool_config_deprecations():` was asserting that the Deprecation warning would be emitted, which is what we want. We are explicitly testing, that our code warns about our deprecated API. Ignoring it will make it so that we wouldn't notice should we, by accident, remove the deprecation (see also [this discussion](https://github.com/neo4j/neo4j-python-driver/pull/832#discussion_r999614790)).

File: tests/unit/test_conf.py [hunk 11]
```diff
@@ -133,11 +135,9 @@ def test_pool_config_consume_key_not_valid():
     test_config["not_valid_key"] = "test"
 
     with pytest.raises(ConfigurationError) as error:
-        # might or might not warn DeprecationWarning, but we're only
-        # interested in the error
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: Why did you remove the comment?

File: tests/unit/test_conf.py [hunk 12]
```diff
@@ -133,11 +135,9 @@ def test_pool_config_consume_key_not_valid():
...
         with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", category=DeprecationWarning)
-            _ = PoolConfig.consume(test_config)
+            warnings.simplefilter("ignore", DeprecationWarning)
+            consumed_pool_config = PoolConfig.consume(test_config)
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: `consumed_pool_config` is an unused variable. That's why it was assigned to `_`. You could also not assign it to anything.

File: tests/unit/test_conf.py [hunk 13]
```diff
@@ -146,7 +146,8 @@ def test_pool_config_set_value():
 
     test_config = dict(test_pool_config)
 
-    with _pool_config_deprecations():
+    with warnings.catch_warnings():
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: See comment above. This changes what we're testing for.

File: tests/unit/test_conf.py [hunk 14]
```diff
@@ -165,37 +166,37 @@ def test_pool_config_set_value():
 def test_pool_config_consume_and_then_consume_again():
 
     test_config = dict(test_pool_config)
-    with _pool_config_deprecations():
+
+    with warnings.catch_warnings():
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: :eyes:

File: tests/unit/test_conf.py [hunk 15]
```diff
@@ -165,37 +166,37 @@ def test_pool_config_set_value():
...
+        warnings.simplefilter("ignore", DeprecationWarning)
         consumed_pool_config = PoolConfig.consume(test_config)
+
     assert consumed_pool_config.encrypted is False
     consumed_pool_config.encrypted = "test"
 
     with pytest.raises(AttributeError):
-        _ = PoolConfig.consume(consumed_pool_config)
+        consumed_pool_config = PoolConfig.consume(consumed_pool_config)
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: unused variable

File: tests/unit/test_conf.py [hunk 16]
```diff
@@ -165,37 +166,37 @@ def test_pool_config_set_value():
...
 
-    with _pool_config_deprecations():
-        consumed_pool_config = PoolConfig.consume(
-            dict(consumed_pool_config.items())
-        )
-    with _pool_config_deprecations():
-        consumed_pool_config = PoolConfig.consume(
-            dict(consumed_pool_config.items())
-        )
+    with warnings.catch_warnings():
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: :eyes:

File: tests/unit/test_conf.py [hunk 17]
```diff
@@ -263,19 +261,98 @@ def test_init_session_config_with_not_valid_key():
...
+    pool_config = PoolConfig.consume(config)
+    assert pool_config.encrypted is True
+    ssl_context = pool_config.get_ssl_context()
+    assert isinstance(ssl_context, ssl.SSLContext)
+    _assert_context_tls_1_2(ssl_context)
+    assert ssl_context.check_hostname is False
+    assert ssl_context.verify_mode is ssl.CERT_NONE
+
+
+def _assert_context_tls_1_2(ctx):
```
Comment (robsdedude) [inline tests/unit/test_conf.py:386]: And no assertion for Python below 3.7?

File: neo4j/conf.py [hunk 18]
```diff
@@ -357,11 +363,9 @@ class SessionConfig(WorkspaceConfig):
 
 class TransactionConfig(Config):
     """ Transaction configuration. This is internal for now.
-
     neo4j.session.begin_transaction
     neo4j.Query
     neo4j.unit_of_work
-
```
Comment (robsdedude) [inline neo4j/conf.py]: Why was it changed?

File: tests/unit/test_conf.py [hunk 19]
```diff
@@ -263,19 +261,98 @@ def test_init_session_config_with_not_valid_key():
...
+    assert ssl_context.check_hostname is False
+    assert ssl_context.verify_mode is ssl.CERT_NONE
+
+
+def _assert_mock_tls_1_2(mock):
+    mock.assert_called_once_with(ssl.PROTOCOL_TLS_CLIENT)
+    if sys.version_info >= (3, 7):
+        assert mock.return_value.minimum_version == ssl.TLSVersion.TLSv1_2
+    else:
+        assert mock.mock_calls[1][1][0] == ssl.OP_NO_TLSv1
```
Comment (robsdedude) [inline tests/unit/test_conf.py]: 1) I find `mock.mock_calls[1][1][0]` hard to understand
2) It's pretty fragile as it asserts that the seconds thing done with the SSLContext is t configuring `ssl.OP_NO_TLSv1`. What if we configure something else before?
3) This only tests if TLS1 is disabled. We also want to ensure that TLS1.1 is disabled.

Here's a suggestion on how it could be done. I tried doing it with mock call assertions but that didn't really pan out because the `__ior__` calls would stack up and you'd still have to assume some configuration order.
```suggestion
@pytest.mark.parametrize("config", (
    {"encrypted": True},
    {"encrypted": True, "trust": TRUST_SYSTEM_CA_SIGNED_CERTIFICATES},
))
def test_trust_system_cas_mock(config, mocker):
    ssl_context_mock = mocker.patch("ssl.SSLContext", autospec=True)
    ssl_context_mock.return_value.options = 0
    pool_config = PoolConfig.consume(config)
    assert pool_config.encrypted is True
    ssl_context = pool_config.get_ssl_context()
    _assert_mock_tls_1_2(ssl_context_mock)
    ssl_context_mock.return_value.load_default_certs.assert_called_once_with()
    ssl_context_mock.return_value.load_verify_locations.assert_not_called()
    assert ssl_context.check_hostname
    assert ssl_context.verify_mode


@pytest.mark.parametrize("config", (
    {"encrypted": True, "trust": TRUST_ALL_CERTIFICATES},
))
def test_trust_all_mock(config, mocker):
    ssl_context_mock = mocker.patch("ssl.SSLContext", autospec=True)
    ssl_context_mock.return_value.options = 0
    pool_config = PoolConfig.consume(config)
    assert pool_config.encrypted is True
    ssl_context = pool_config.get_ssl_context()
    _assert_mock_tls_1_2(ssl_context_mock)
    ssl_context_mock.return_value.load_verify_locations.assert_not_called()
    assert ssl_context.check_hostname is False
    assert ssl_context.verify_mode is ssl.CERT_NONE


def _assert_mock_tls_1_2(mock):
    mock.assert_called_once_with(ssl.PROTOCOL_TLS_CLIENT)
    if sys.version_info >= (3, 7):
        assert mock.return_value.minimum_version == ssl.TLSVersion.TLSv1_2
    else:
        assert mock.return_value.options & ssl.OP_NO_TLSv1
        assert mock.return_value.options & ssl.OP_NO_TLSv1_1
        assert not mock.return_value.options & ssl.OP_NO_TLSv1_2
        assert not mock.return_value.options & ssl.OP_NO_TLSv1_3
```

File: tests/unit/test_conf.py [hunk 20]
```diff
@@ -182,20 +184,30 @@ def test_pool_config_consume_and_then_consume_again():
...
 
+    with _pool_config_deprecations():
+        consumed_pool_config = PoolConfig.consume(
+            dict(consumed_pool_config.items())
+        )
+    with _pool_config_deprecations():
+        consumed_pool_config = PoolConfig.consume(
+            dict(consumed_pool_config.items())
+        )
+
```
Comment (robsdedude) [inline tests/unit/test_conf.py]:
```suggestion
```

File: tests/unit/test_conf.py [hunk 21]
```diff
@@ -182,20 +184,30 @@ def test_pool_config_consume_and_then_consume_again():
...
     assert consumed_pool_config.encrypted == "test"
 
 
 def test_config_consume_chain():
 
     test_config = {}
+
     test_config.update(test_pool_config)
+
```
Comment (robsdedude) [inline tests/unit/test_conf.py]:
```suggestion
    test_config.update(test_pool_config)
```

File: tests/unit/test_conf.py [hunk 22]
```diff
@@ -251,7 +264,11 @@ def test_init_session_config_with_not_valid_key():
     test_config_a = {"connection_acquisition_timeout": 111}
     workspace_config = WorkspaceConfig.consume(test_config_a)
 
-    test_config_b = {"default_access_mode": READ_ACCESS, "connection_acquisition_timeout": 333, "not_valid_key": None}
+    test_config_b = {
+        "default_access_mode": READ_ACCESS,
+        "connection_acquisition_timeout": 333,
+        "not_valid_key": None
```
Comment (robsdedude) [inline tests/unit/test_conf.py]:
```suggestion
        "not_valid_key": None,
```

**Review by robsdedude — APPROVED** [review]
Changing the CI pipelines to run TestKit tests with Py36 and Py310 (instead of 36 and 39) before merging.

**Review by bigmontz — APPROVED** [review]
🎃
